plugin/client_rpc.go

```diff
@@ -720,3 +720,54 @@ func (s *apiRPCServer) LogError(args *Z_LogErrorArgs, returns *Z_LogErrorReturns
 	}
 	return nil
 }
+
+type Z_InstallPluginArgs struct {
+	PluginStreamID uint32
+	B              bool
+}
+
+type Z_InstallPluginReturns struct {
+	A *model.Manifest
+	B *model.AppError
+}
+
+func (g *apiRPCClient) InstallPlugin(file io.Reader, replace bool) (*model.Manifest, *model.AppError) {
+	pluginStreamID := g.muxBroker.NextId()
+
+	go func() {
+		uploadPluginConnection, err := g.muxBroker.Accept(pluginStreamID)
+		if err != nil {
+			log.Print("Plugin failed to upload plugin. MuxBroker could not Accept connection", mlog.Err(err))
+			return
+		}
+		defer uploadPluginConnection.Close()
+		serveIOReader(file, uploadPluginConnection)
+	}()
+
+	_args := &Z_InstallPluginArgs{pluginStreamID, replace}
+	_returns := &Z_InstallPluginReturns{}
+	if err := g.client.Call("Plugin.InstallPlugin", _args, _returns); err != nil {
+		log.Print("RPC call InstallPlugin to plugin failed.", mlog.Err(err))
+	}
+
+	return _returns.A, _returns.B
+}
+
+func (g *apiRPCServer) InstallPlugin(args *Z_InstallPluginArgs, returns *Z_InstallPluginReturns) error {
+	receivePluginConnection, err := g.muxBroker.Dial(args.PluginStreamID)
```

Comment:
Do yo mean why starting to `Accept()` at the client side instead of doing it in the reverse way? Please correct me if I misunderstood your question. 

I think it's not possible to `Accept()` from the server side without doing some sort of synchronization or other tricks because client won't have any information about if server is ready to accept connections.

```
> [T] app calls: client.InstallPlugin()
  > [T] client.InstallPlugin() calls server.InstallPlugin() with stream-id.
    > [T(+1)]-(network is slower) server receives the call, starts to `Accept()` connections for stream-id.
  > [T(+0.001)]-(goroutine is faster) client tries to `go Dial()` with stream-id but it's most likely server has not started to Accept() yet.
  > [T(+0.5)] client cannot Dial() and thus, cannot send data to server and returns with error.
```

To overcome this, client needs to know when server is ready to `Accept()`. The easiest way would be pooling the `err := Dial()` until `err == nil` but this solution is not elegant.  

--
Although multiplexed connections works in both ways and I think it's common to initialize a connection from both ends since it really depends on scenario.

Also, in the current code, once client starts to accept connections with `Accept()` but if server is unable to `Dial()` for some reason like a network failure during the _RPC_ call initiative from the client _client.RPC("Plugin.InstallPlugin")_, still `Accept()` will be cancelled after [this timeout hits](https://github.com/hashicorp/go-plugin/blob/2543bf2/mux_broker.go#L186) and there will be nothing left out in the memory.

Reply:
Actually, I'm just wondering if we can test for the `InstallPlugin` to be implemented before setup the stream only they to have to immediately close it:
```go
func (g *apiRPCServer) InstallPlugin(args *Z_InstallPluginArgs, returns *Z_InstallPluginReturns) error {
	hook, ok := g.impl.(interface {
 		InstallPlugin(file io.Reader, replace bool) (*model.Manifest, *model.AppError)
 	}); 
        if !ok {
 		return encodableError(fmt.Errorf("API InstallPlugin called but not implemented."))
        }

	receivePluginConnection, err := g.muxBroker.Dial(args.PluginStreamID)
        // ...

        returns.A, returns.B = hook.InstallPlugin(pluginReader, args.B)
        return nil
}
```

Reply:
Oh, of course. I see it know, we should call `Dial()` if there is an implementation for `InstallPlugin ()`.